In [1]:
from src.agents.offer_agent.agent import *

Status:  Initialising nodes
Status:  Initialising ChatOpenAI
Status:  ChatOpenAI initialised
Status:  Initialising nodes
Status:  Initialising ChatOpenAI
Status:  ChatOpenAI initialised


In [2]:
segments = []
metrics = ['1', '2', '3']
ideas = []
human_remark = """Crit Fishermen (Fishing + Archery Meme Builds)

Idea: “Trickshot Fishing Contests”
Periodically host an event where fish spawns are temporarily replaced by tricky “crit-fish” mobs that jump out of water. Players use bows to try and snipe these flying fish for rare loot (e.g., high-luck items). This merges the Fishing skill’s Luck attribute with Archery’s crit mechanics, embracing the meme-worthy synergy.
"""

In [3]:
#offers = get_offers(metric_ids=metrics, segment_ids=segments, human_remark=human_remark, idea_ids=ideas)

In [4]:
#print(offers['offers'])

In [5]:
#chat_id = register_new_chat(offers)

In [6]:
#chat_id

In [7]:
player_segments_dict = {
    # Skill Specialists
    "Skill Min-Maxers": {
        "segment_name": "Skill Min-Maxers",
        "segment_description": (
            "Obsessively optimize specific skills (e.g., Alchemy-only players) to exploit "
            "stat synergies."
        ),
        "ideas": [
            (
                "Alchemical Master Trials: Add a series of solo Alchemy-oriented trials located "
                "in a hidden 'Arcane Lab.' Players must craft specialized potions under strict time "
                "limits to unlock advanced alchemical recipes. This caters to min-maxers’ desire to "
                "push one skill to its limit and offers them exclusive potions not available elsewhere."
            )
        ]
    },
    "Jack-of-All-Trades": {
        "segment_name": "Jack-of-All-Trades",
        "segment_description": (
            "Balance skill progression to avoid triggering adaptive mob scaling too quickly."
        ),
        "ideas": [
            (
                "Balance Badges: Introduce 'Balance Badges' earned by keeping all skills within "
                "a small level range (e.g., no skill more than 5 levels apart). Collecting enough "
                "badges grants small global buffs and cosmetic titles, rewarding a well-rounded approach "
                "without accelerating difficulty scaling."
            )
        ]
    },
    "Lone Enchanters": {
        "segment_name": "Lone Enchanters",
        "segment_description": (
            "Focus solely on Enchanting/Wisdom to dominate the magic-resistance meta."
        ),
        "ideas": [
            (
                "Enchanter’s Sanctum: Create a special Sanctuary biome where a magical barrier "
                "drains health unless you have high Enchanting/Wisdom. Inside, players find powerful "
                "enchanting tomes that grant unique enchantments, rewarding those who invest heavily in "
                "magical mastery."
            )
        ]
    },
    "Crit Fishermen": {
        "segment_name": "Crit Fishermen",
        "segment_description": (
            "Combine Fishing (Luck) and Archery (Crit) for meme-y one-shot builds."
        ),
        "ideas": [
            (
                "Trickshot Fishing Contests: Host events where fish spawn as 'crit-fish' that jump "
                "out of water. Players use bows to snipe these flying fish, earning rare luck-based items "
                "or high-crit gear. This blends Fishing’s Luck with Archery’s crit mechanics for a fun, "
                "gimmicky synergy."
            )
        ]
    },

    # Combat Playstyles
    "Blunt Brawlers": {
        "segment_name": "Blunt Brawlers",
        "segment_description": (
            "Prefer fast-paced, low-damage weapons for chaotic PvP skirmishes."
        ),
        "ideas": [
            (
                "Knockout Colosseum: Create a PvP arena mode where only blunt weapons are allowed. "
                "Winners earn 'Brawler Tokens' that can be exchanged for gear improving stun duration or "
                "reducing stamina costs. This sub-mode focuses on chaotic, rapid-fire melee combat."
            )
        ]
    },
    "Firearm Hoarders": {
        "segment_name": "Firearm Hoarders",
        "segment_description": (
            "Obsessively collect ammo to unleash rare guns during End War climaxes."
        ),
        "ideas": [
            (
                "Ammo Auction Week: Once per season, a server-wide event spotlights special ammo "
                "types in auctions or high-tier mob drops. Hoarders can strategize to corner the ammo "
                "market or barter with others, fueling a dynamic economy and enabling massive End War "
                "firearm showdowns."
            )
        ]
    },
    "Tanky Foragers": {
        "segment_name": "Tanky Foragers",
        "segment_description": (
            "Stack Toughness via Foraging/Defense to face-tank elite mobs."
        ),
        "ideas": [
            (
                "The Great Forage Siege: Occasionally spawn wave-based siege events in heavily "
                "forested areas. Surviving each wave grants valuable resources or special 'bark armor' "
                "with high defense. Tanky Foragers excel, leveraging foraged items that boost "
                "Toughness."
            )
        ]
    },
    "Hit-and-Run Archers": {
        "segment_name": "Hit-and-Run Archers",
        "segment_description": (
            "Abuse Agility/Evasion to kite enemies in open-world PvP."
        ),
        "ideas": [
            (
                "Open-Range Ambushes: Randomly generate NPC caravans carrying high-tier loot. "
                "Hit-and-Run Archers excel at kiting caravan guards, using agility for quick strikes. "
                "Daily or weekly caravan route announcements encourage active scouting and nimble PvP."
            )
        ]
    },

    # Economy & Social
    "Black Market Bandits": {
        "segment_name": "Black Market Bandits",
        "segment_description": (
            "Trade illegal resources during secret events and sabotage auctions."
        ),
        "ideas": [
            (
                "Hidden Black Market Hubs: Vaguely announced pop-up markets appear in random locations. "
                "Bandits can trade contraband or disrupt rival deals. This fosters a secretive, high-stakes "
                "economic subculture revolving around illicit goods."
            )
        ]
    },
    "Land Barons": {
        "segment_name": "Land Barons",
        "segment_description": (
            "Monetize claim blocks by leasing secured zones to casual players."
        ),
        "ideas": [
            (
                "Real Estate Broker System: Let experienced players buy 'Broker Licenses' to subdivide "
                "large claims into rentable plots. They can set rent fees and building rules, creating a "
                "player-driven property market and ongoing passive income for Land Barons."
            )
        ]
    },
    "War Profiteers": {
        "segment_name": "War Profiteers",
        "segment_description": (
            "Stockpile End War-critical resources to inflate prices mid-season."
        ),
        "ideas": [
            (
                "Critical Resource Index: Display a live ticker of key items (e.g., gunpowder, Nether shards) "
                "that become vital near the End War. Prices rise with demand, enabling profiteers to exploit "
                "scarcity—but risking bounties or sabotage if they overcharge."
            )
        ]
    },
    "Guild Loan Sharks": {
        "segment_name": "Guild Loan Sharks",
        "segment_description": (
            "Lend gear/currency to newbies in exchange for End War conscription."
        ),
        "ideas": [
            (
                "Debt Contract System: Officially allow guilds to post 'Debt Contracts' on a bulletin board, "
                "offering gear or currency with the condition that borrowers fight in the End War under the "
                "guild banner. Failure to repay triggers bounties or resource lockouts, adding real consequences."
            )
        ]
    },

    # Event-Driven
    "Nether Nomads": {
        "segment_name": "Nether Nomads",
        "segment_description": (
            "Rush Nether events to farm rare mats before others unlock the zone."
        ),
        "ideas": [
            (
                "Nether Expeditions: Time-limited global events periodically open the Nether early. "
                "Nomads who dash in can gather exclusive resources or battle mini-bosses, but risk "
                "facing stronger Overworld mobs on re-entry due to lingering Nether energy."
            )
        ]
    },
    "Arena Addicts": {
        "segment_name": "Arena Addicts",
        "segment_description": (
            "Ignore survival mechanics to live purely for ranked PvP leaderboards."
        ),
        "ideas": [
            (
                "Arena Seasons: Run a dedicated Ranked PvP season in parallel with main survival. "
                "Arena-focused players climb leaderboards for exclusive vanity titles and top-tier gear, "
                "keeping them engaged without overshadowing other server aspects."
            )
        ]
    },
    "Event Tourists": {
        "segment_name": "Event Tourists",
        "segment_description": (
            "Log in only during world events for FOMO-driven loot grabs."
        ),
        "ideas": [
            (
                "Seasonal Passports: Each event gives a collectible 'stamp' in a passport item. "
                "Accumulate enough stamps for a big end-of-season reward (cosmetics, unique gear). "
                "Perfectly aligns with FOMO-driven playstyles."
            )
        ]
    },
    "Boss Baiters": {
        "segment_name": "Boss Baiters",
        "segment_description": (
            "Lure event-only mobs into rival territories to wreak havoc."
        ),
        "ideas": [
            (
                "Mob Relay Challenge: Event mobs drop double loot if defeated far from spawn. "
                "Boss Baiters exploit this by dragging powerful enemies across the map—often into "
                "enemies’ bases—courting high risks and high rewards."
            )
        ]
    },

    # Survival Strategists
    "Pacifist Homesteaders": {
        "segment_name": "Pacifist Homesteaders",
        "segment_description": (
            "Avoid combat entirely by maxing Farming/Claiming early."
        ),
        "ideas": [
            (
                "Farming Diplomacy Contracts: Implement a feature to 'hire' combat players for "
                "base defense in exchange for crops or claim blocks. A scoreboard tracks successful "
                "pacifist-run homesteads, rewarding peaceful progress and alliances."
            )
        ]
    },
    "Regen Rats": {
        "segment_name": "Regen Rats",
        "segment_description": (
            "Stack Regeneration/Fighting stats to outheal enemies in prolonged wars."
        ),
        "ideas": [
            (
                "Regeneration Sprint Trials: Weekly 'Endurance Test' events spawn waves of moderate-damage mobs. "
                "The longest survivor wins special gear boosting regen further. Regen Rats thrive, turning "
                "their healing builds into a recognized server challenge."
            )
        ]
    },
    "Nightmare Mode Masochists": {
        "segment_name": "Nightmare Mode Masochists",
        "segment_description": (
            "Deliberately overlevel skills to spawn elite mobs in starter zones."
        ),
        "ideas": [
            (
                "Overleveler’s Peril: Create a hardcore 'Nightmare' toggle that spawns mobs scaled to the "
                "player’s highest skill everywhere they go. Completing 'Nightmare Achievements' for surviving "
                "in low-level zones at high skill grants special bragging-rights cosmetics or land expansions."
            )
        ]
    },
    "Seed Theorists": {
        "segment_name": "Seed Theorists",
        "segment_description": (
            "Spend hours rerolling worlds for perfect procedural-gen base locations."
        ),
        "ideas": [
            (
                "Custom Seed Competitions: Each month, showcase top player-submitted seeds on a rotating sub-server. "
                "Winning seeds earn the submitter in-game currency, naming rights, or unique items. Perfect for "
                "players who obsess over ideal world generation."
            )
        ]
    },

    # End War Focus
    "Sleeper Agents": {
        "segment_name": "Sleeper Agents",
        "segment_description": (
            "Hide power-leveling until the final week to dominate End War rankings."
        ),
        "ideas": [
            (
                "Stealth Growth Tracker: An optional mode hides your skill gains from public leaderboards. "
                "Once you disable it near the End War, you gain a temporary 'Shadow Ascendant' buff, reflecting "
                "your sudden reveal of massive power."
            )
        ]
    },
    "Sacrificial Lambs": {
        "segment_name": "Sacrificial Lambs",
        "segment_description": (
            "Spec into suicide builds (e.g., explosive alchemy) to disrupt enemy teams."
        ),
        "ideas": [
            (
                "Kamikaze Contracts: Guilds post 'Kamikaze Contracts' rewarding players who use high-risk, "
                "self-destructive tactics in End War. Each contract logs kills or chaos caused, granting points "
                "redeemable post-war for rare gear or currency."
            )
        ]
    },
    "War Archivists": {
        "segment_name": "War Archivists",
        "segment_description": (
            "Document every End War outcome to predict future meta shifts."
        ),
        "ideas": [
            (
                "Historical Battle Chronicle: Post-End War, a detailed log or timeline shows kill counts, alliances, "
                "and winning strategies. Archivists can purchase 'Analysis Tools' to parse past data for insights, "
                "fueling their quest to forecast the next war’s meta."
            )
        ]
    },
    "Mercenary Carry": {
        "segment_name": "Mercenary Carry",
        "segment_description": (
            "Sell End War victory services in exchange for rare blueprints."
        ),
        "ideas": [
            (
                "Mercenary License & Marketplace: High-level players list 'carry services' in an official marketplace. "
                "Clients pay with rare items or currency upfront. Successful carries earn the Merc a positive reputation "
                "and exclusive 'Champion for Hire' cosmetic titles, fostering a professional in-game mercenary culture."
            )
        ]
    },

    # Quirky/Roleplay
    "Lore LARPers": {
        "segment_name": "Lore LARPers",
        "segment_description": (
            "Invent backstories and in-game factions despite no official lore."
        ),
        "ideas": [
            (
                "Storyteller Module & Faction Halls: Allow players to build 'Faction Halls' and create custom NPCs, "
                "interactive signs, or small questlines. This fosters an in-game roleplay scene and encourages "
                "creative narratives, despite the sandbox nature."
            )
        ]
    },
    "Drip Cultivators": {
        "segment_name": "Drip Cultivators",
        "segment_description": (
            "Prioritize cosmetic base designs over defensive efficiency."
        ),
        "ideas": [
            (
                "Architecture Showcase & Style Points: A monthly building contest judged by staff or players. "
                "Winners earn 'Style Points' redeemable for exclusive decorative blocks, colored lighting, or "
                "vanity items, catering to those who care about aesthetics above all else."
            )
        ]
    },
    "Pet Hoarders": {
        "segment_name": "Pet Hoarders",
        "segment_description": (
            "Collect mobs (named/caged) as trophies instead of killing them."
        ),
        "ideas": [
            (
                "Menagerie System & Pet Exhibitions: Let players register and display captured mobs "
                "in specialized zoo-like areas. Monthly exhibitions give titles ('Beast Tamer') and unique "
                "rewards for the most exotic or numerous collections."
            )
        ]
    },
    "Luck Zealots": {
        "segment_name": "Luck Zealots",
        "segment_description": (
            "Worship RNG by maxing Luck stats and gambling all resources."
        ),
        "ideas": [
            (
                "Luck Shrine & Gambling Dungeons: A 'Luck Shrine' grants daily RNG-based buffs or debuffs. "
                "Pair it with procedurally generated 'Gambling Dungeons' full of random-tier loot chests. "
                "Luck-focused players literally gamble progression each run."
            )
        ]
    }
}


In [21]:
new_segments_dict = {
    31: {
        "segment_name": "Nomadic Alchemists",
        "segment_description": (
            "Wander the map in search of rare ingredients and hidden labs to perfect unconventional potions."
        ),
        "ideas": [
            "Hidden Ingredient Challenges: Scatter ultra-rare flowers, fungi, or crystals in remote biomes. Nomadic Alchemists who collect these can brew specialized potions (e.g., invisibility-from-mobs or extra-long regeneration) that are nearly impossible to craft with standard materials."
        ]
    },
    32: {
        "segment_name": "Siege Engineers",
        "segment_description": (
            "Obsess over building advanced war machines and fortified structures to excel in large-scale PvP sieges."
        ),
        "ideas": [
            "Fortification Contracts: Allow players to sell or rent siege contraptions—like TNT cannons or auto-crossbow turrets—to factions gearing up for End War. This fosters a sub-economy around constructing advanced defenses and offensive platforms."
        ]
    },
    33: {
        "segment_name": "Aerial Adventurers",
        "segment_description": (
            "Use high Agility stats and special glide or flight mechanics (e.g., Elytra-like items) to explore sky biomes."
        ),
        "ideas": [
            "Sky Island Expeditions: Procedurally generate floating islands with rare resources or hidden dungeons. Aerial Adventurers can glide between them, collecting unique loot like 'Wind Gemstones' used for ultra-light gear or decorative building blocks."
        ]
    },
    34: {
        "segment_name": "Dungeon Speedrunners",
        "segment_description": (
            "Try to complete dungeons and raids as fast as possible, often ignoring loot to set record times."
        ),
        "ideas": [
            "Timed Dungeon Challenges: Implement official leaderboards that track players’ clear times for each dungeon. Offer exclusive cosmetic rewards (speedrunner capes, special titles) for topping weekly or monthly charts."
        ]
    },
    35: {
        "segment_name": "Harvest Hustlers",
        "segment_description": (
            "Minimize combat and instead focus on building large-scale farms to dominate the resource economy."
        ),
        "ideas": [
            "Crop Co-op Auctions: A specialized auction system for bulk farming produce. Harvest Hustlers can set up 'crop shares' where investors earn dividends in the form of produce or in-game currency, boosting cooperative agriculture."
        ]
    },
    36: {
        "segment_name": "Treasure Cartographers",
        "segment_description": (
            "Obsess over mapping and discovering every secret zone, labyrinth, or rare chest location."
        ),
        "ideas": [
            "Expedition Logbooks: Players can craft logbooks to record discovered points of interest, secret chests, and puzzle rooms. Selling completed logbooks to others fosters a cartography-based trade network."
        ]
    },
    37: {
        "segment_name": "Blood Moon Cultists",
        "segment_description": (
            "Deliberately seek out the most dangerous nights (blood moons) to hunt elite mobs for high-tier drops."
        ),
        "ideas": [
            "Blood Moon Beacon: A craftable item that guarantees the next night spawns a 'Blood Moon' event for that chunk. Cultists use it to force intense hunts, yielding exclusive red-hued gear or dark enchant materials."
        ]
    },
    38: {
        "segment_name": "Food Critic Farmers",
        "segment_description": (
            "Curate the perfect array of crops, livestock, and cooking recipes to maximize culinary bonuses."
        ),
        "ideas": [
            "Gourmet Cooking Showdowns: Host periodic cooking contests where participants present unique dishes (boosting stats like health, luck, or regeneration). Judges (NPC or player-voted) award rare cooking utensils or 'Master Chef' titles."
        ]
    },
    39: {
        "segment_name": "Arcane Archivists",
        "segment_description": (
            "Collect every known enchantment or magical relic, aiming to complete a 'living library' of arcane knowledge."
        ),
        "ideas": [
            "Enchanted Library Blueprint: A special buildable structure that reveals illusions of missing enchantments and hints on where to find them. Completing the library offers a global buff or unique cosmetic effect."
        ]
    },
    40: {
        "segment_name": "Eco Guardians",
        "segment_description": (
            "Preserve biomes by preventing over-farming, replanting resources, and policing polluters."
        ),
        "ideas": [
            "Biome Restoration Quests: Introduce timed quests where entire forests or coral reefs are threatened by over-mining/fishing. Eco Guardians who restore or protect these areas get eco-themed cosmetics and trait buffs like 'Nature’s Protection.'"
        ]
    },
    41: {
        "segment_name": "Night Vision Hunters",
        "segment_description": (
            "Embrace darkness as an ally. Equip night-vision gear and hunt nocturnal mobs for rare loot or XP."
        ),
        "ideas": [
            "Nightfall Hunts: Add special late-night bosses or 'Stalker mobs' that drop glow-infused armor shards. Night Vision Hunters excel here, turning darkness into a tactical advantage in stealth-based hunts."
        ]
    },
    42: {
        "segment_name": "Mob Tamer Guild",
        "segment_description": (
            "Domesticate hostile mobs with specialized items or potions, training them as battlefield allies."
        ),
        "ideas": [
            "Taming Trials: Create unique 'Bonding Items' that allow advanced players to tame certain hostile mobs. Tamed mobs can guard bases or follow their owner, adding complexity to both PvE and PvP. The server can host 'Best Tamed Mob' showcases."
        ]
    },
    43: {
        "segment_name": "Underworld Divers",
        "segment_description": (
            "Exploit advanced Mining and Foraging to thrive in deep cave networks, rarely surfacing."
        ),
        "ideas": [
            "Crystal Cavern Expeditions: Populate deeper cave layers with luminous crystals that boost certain skills or serve as high-value trade goods. Underworld Divers become specialists at navigating these labyrinthine depths for profit."
        ]
    },
    44: {
        "segment_name": "Toxic Tricksters",
        "segment_description": (
            "Max out Alchemy to craft potent poisons and sabotage items, thriving on subterfuge in PvP."
        ),
        "ideas": [
            "Poison Mastery: Add advanced poison recipes that can be slipped into enemy item trades or sprinkled into farmland. Debuffs range from mild slowness to lethal damage over time, fueling cunning warfare."
        ]
    },
    45: {
        "segment_name": "Curse-Forgers",
        "segment_description": (
            "Enchant items with powerful but risky negative effects (e.g., cursed swords that boost damage but drain health)."
        ),
        "ideas": [
            "Hex Anvils: Rare crafting stations found in cursed biomes. Using them can apply dual-enchantments: big offensive buffs with nasty side effects. Rewards expert knowledge and risk-taking builds."
        ]
    },
    46: {
        "segment_name": "Tribalistic Roleplayers",
        "segment_description": (
            "Form small, lore-based tribes or clans, living in thematically consistent areas (e.g., desert tribes)."
        ),
        "ideas": [
            "Tribal Lore Tokens: Each tribe can create unique lore tokens, gained by performing rituals or hunts tied to their biome. Collecting a full set bestows a 'Tribal Spirit' buff for group-based synergy in raids or hunts."
        ]
    },
    47: {
        "segment_name": "Wasteland Scrappers",
        "segment_description": (
            "Thrive in resource-scarce biomes, turning junk or scrap into makeshift gear and cunning traps."
        ),
        "ideas": [
            "Scrapyard Upgrades: Desert or wasteland biomes feature 'scrap piles' for raw materials. Wasteland Scrappers combine them to craft unique contraptions (e.g., wire traps, spiked barricades), excelling in a post-apocalyptic style."
        ]
    },
    48: {
        "segment_name": "Redstone Savants",
        "segment_description": (
            "Expert at building complex redstone contraptions to automate farming, base defense, or resource processing."
        ),
        "ideas": [
            "Engineer’s Guild: A dedicated crafting tree for advanced redstone components (e.g., wireless signals, portable turrets). Completing specific engineering milestones grants badges and specialized craft recipes for high-level automation."
        ]
    },
    49: {
        "segment_name": "Spiritual Monks",
        "segment_description": (
            "Rely on minimal gear, high wisdom (Enchanting), and disciplined self-imposed rules (e.g., no metal weapons)."
        ),
        "ideas": [
            "Monastic Pilgrimages: A curated path of temple-like structures scattered across the world. Each temple bestows a small permanent stat buff if the pilgrim meets certain humility trials—like disarming or offering rare resources to shrine NPCs."
        ]
    },
    50: {
        "segment_name": "Bargain Barters",
        "segment_description": (
            "Negotiate and trade relentlessly for the best deals, often flipping items for massive in-game profits."
        ),
        "ideas": [
            "Trade Fair Events: Host monthly trade festivals where players set up stalls in a neutral zone. Bargain Barters excel here, hunting for undervalued items and forging profitable deals. Exclusive 'Master Trader' titles reward the best deals made."
        ]
    },
    51: {
        "segment_name": "Corrupted Pathfinders",
        "segment_description": (
            "Intentionally allow themselves to be ‘corrupted’ by dark forces to unlock forbidden skill bonuses."
        ),
        "ideas": [
            "Nightmare Shrines: Players must sacrifice certain stats or items to gain potent but risky 'Corrupted Buffs' (e.g., +20% damage but periodic health drains). They become walking threats, altering how the community interacts with them."
        ]
    },
    52: {
        "segment_name": "Mobile Fortress Builders",
        "segment_description": (
            "Design movable bases or caravans loaded with defensive turrets, preferring a life on the road."
        ),
        "ideas": [
            "Caravan Crafting Bench: A specialized build station to create mobile structures that can be placed on rails or slow-moving contraptions. Perfect for roving from biome to biome while retaining a defensible 'home on wheels.'"
        ]
    },
    53: {
        "segment_name": "Cursed Treasure Hunters",
        "segment_description": (
            "Ignore the danger of cursed dungeons or traps to score legendary loot, always chasing the next big find."
        ),
        "ideas": [
            "Trap-Laden Vaults: Generate extremely hazardous vaults guarded by advanced puzzles or illusions. The final chest yields 'Heirloom Relics' that provide unique powers at the cost of potential curses. Ideal for adrenaline-seeking treasure fiends."
        ]
    },
    54: {
        "segment_name": "Silver-Tongued Diplomats",
        "segment_description": (
            "Foster alliances, negotiate treaties, and de-escalate conflicts, focusing on politics rather than combat."
        ),
        "ideas": [
            "Treaty System: Implement a UI where players or guilds can propose formal alliances or truces with conditions (resource tributes, restricted expansions). Diplomats gain rep-based perks (like cheaper trading fees or small skill boosts)."
        ]
    },
    55: {
        "segment_name": "Riftwalkers",
        "segment_description": (
            "Exploit dimensional portals (Nether, End, or special 'rift' events) to chain-hop worlds for new content."
        ),
        "ideas": [
            "Rift Challenges: Timed multi-stage events requiring players to pass through sequential portals, each with a distinct biome/mini-boss. Completion yields cross-dimensional gear that offers synergy across different realms."
        ]
    },
    56: {
        "segment_name": "Aquatic Overlords",
        "segment_description": (
            "Live predominantly underwater, constructing bubble cities and farming aquatic mobs or resources."
        ),
        "ideas": [
            "Deep-Sea Domes: Unlock craftable 'aqua-domes' that protect entire underwater enclaves from drowning or water pressure. Aquatic Overlords trade specialized fish, coral, and pearl-based building materials on a niche but lucrative market."
        ]
    },
    57: {
        "segment_name": "Minion Masters",
        "segment_description": (
            "Use advanced summoning or breeding to spawn lesser creatures that assist in gathering or combat."
        ),
        "ideas": [
            "Summoning Circles: A skill-based structure allowing players to conjure short-lived minions (harvest drones, battle imps) with resource or XP costs. Minion Masters revolve entire builds around controlling these swarms."
        ]
    },
    58: {
        "segment_name": "Curiosity Collectors",
        "segment_description": (
            "Gather bizarre, novelty items—like offbeat furniture, rare paintings, or developer easter eggs."
        ),
        "ideas": [
            "Museum Showcases: Let players create public “Museums” in claimed land. Admission fees (optional) and item displays featuring ultra-rare curios attract visitors, turning these museums into social hubs."
        ]
    },
    59: {
        "segment_name": "Mythical Beast Riders",
        "segment_description": (
            "Focus on taming or breeding large mountable creatures (e.g., dragons, giant spiders) for travel or war."
        ),
        "ideas": [
            "Sky or Siege Mounts: High-tier breeding systems produce specialized mounts—flying beasts for exploration or heavily armored siege creatures for battles. Taming them requires exclusive items found only in epic raids, making it a true endgame pursuit."
        ]
    },
    60: {
        "segment_name": "Remote Watchers",
        "segment_description": (
            "Monitor the world from secure vantage points, using spyglasses or camera-like contraptions to gather intel."
        ),
        "ideas": [
            "Scrying Mirror Networks: Craft magical devices placed around the map to remotely view key locations. Remote Watchers become invaluable scouts for large factions, selling intel on enemy movement or resource hotspots."
        ]
    },
    61: {
        "segment_name": "Grand Librarians",
        "segment_description": (
            "Store and annotate every discovered recipe, blueprint, or skill synergy in vast libraries for public use."
        ),
        "ideas": [
            "Knowledge Repository: A specialized building that updates with every new recipe or discovered synergy. Librarians can charge players for 'reading rights' or special copies of advanced crafting manuals, forming an academic hub."
        ]
    },
    62: {
        "segment_name": "Minefield Artisans",
        "segment_description": (
            "Expert in layering traps and environmental hazards underground to secure their territory."
        ),
        "ideas": [
            "Subterranean Booby-Traps: An advanced crafting line of pressure mines, fall traps, or gas vents that trigger on intruders. Skilled Artisans can camouflage these hazards so well that only top-tier Defense or Agility players can detect them."
        ]
    },
    63: {
        "segment_name": "Elemental Warriors",
        "segment_description": (
            "Imbue weapons and armor with elemental powers (fire, ice, lightning), building skill sets around specific elements."
        ),
        "ideas": [
            "Elemental Forge Quests: A rotating weekly quest that aligns with a specific element. Completing it grants an 'Element Key' used to infuse gear with unique elemental buffs (e.g., a lightning sword that stuns or a frost bow that slows targets)."
        ]
    },
    64: {
        "segment_name": "Dragonseekers",
        "segment_description": (
            "Single-mindedly pursue legendary dragons or mythical bosses, ignoring standard progression routes."
        ),
        "ideas": [
            "Draconic Sightings: Periodically announce sightings of an ancient dragon that roams the world. Dragonseekers track it using special compasses. Defeating or taming this creature grants unique endgame gear or ephemeral flight abilities."
        ]
    },
    65: {
        "segment_name": "Harvest Festival Enthusiasts",
        "segment_description": (
            "Live for periodic celebrations of growth, where they partake in minigames and flamboyant parades."
        ),
        "ideas": [
            "Seasonal Festivals: Each in-game season ends with a big festival featuring minigames (pumpkin launching, best livestock contest). Enthusiasts focus on prepping for these events, collecting festival tokens exchangeable for festive cosmetics."
        ]
    },
    66: {
        "segment_name": "Arcane Executioners",
        "segment_description": (
            "Combine high Fighting skill with Enchanting to deliver lethal magical attacks in close-range combat."
        ),
        "ideas": [
            "Mana-Infused Blades: A new class of melee weapons that channel enchantments directly into finishing moves. Arcane Executioners can craft enchant combos—like a flame-lash or a void cleave—for devastating results in both PvP and boss fights."
        ]
    },
    67: {
        "segment_name": "Runic Linguists",
        "segment_description": (
            "Decode hidden runes scattered across the land, unveiling secret spells or story lore expansions."
        ),
        "ideas": [
            "Rune-Cracking Puzzles: Place cryptic rune stones in remote areas, each requiring puzzle-solving or specialized stats. Completing an entire rune set reveals advanced spells or a hidden storyline questline that shapes server lore."
        ]
    },
    68: {
        "segment_name": "Pest Control Experts",
        "segment_description": (
            "Eliminate invasive or nuisance mobs (like swarms of small creatures) for bounties and community appreciation."
        ),
        "ideas": [
            "Infestation Events: Random outbreaks of vermin-like creatures or insect swarms threaten farmland. Pest Control Experts are rewarded with bonus currency or skill XP for quickly containing these pests and saving local resources."
        ]
    },
    69: {
        "segment_name": "Land Hoarders",
        "segment_description": (
            "Claim massive swaths of land not just to rent but to prevent others from building, forming 'private empires.'"
        ),
        "ideas": [
            "Territorial Upkeep System: Large landholdings require upkeep in the form of resources or currency. Land Hoarders thrive if they manage resource flows properly, but risk losing claims to 'land decay' if they over-extend."
        ]
    },
    70: {
        "segment_name": "Canyon Racers",
        "segment_description": (
            "Carve racetracks through canyons, using speed-boost potions or specialized agility builds for adrenaline-pumping competitions."
        ),
        "ideas": [
            "Grand Canyon Rally: A server-sponsored race event in a carved-out canyon circuit. Racers use speed potions, jump pads, or gliders. Winners earn 'Speed Demon' cosmetics and a chunk of entry fees in currency."
        ]
    },
    71: {
        "segment_name": "Merchant Marines",
        "segment_description": (
            "Maintain trade routes across vast oceans or rivers, protecting cargo from pirates and aquatic mobs."
        ),
        "ideas": [
            "Naval Convoy System: Let players form fleets of cargo ships with hireable NPC or player 'marines' for protection. Merchant Marines gain maritime perks, like safe harbors or reduced shipping taxes, fueling robust aquatic commerce."
        ]
    },
    72: {
        "segment_name": "Enchanting Purists",
        "segment_description": (
            "Refuse to mix mundane gear with advanced enchantments, pursuing only the highest-tier, single-purpose items."
        ),
        "ideas": [
            "Mono-Enhance Mode: A specialized crafting rule limiting a piece of gear to one 'pure' enchant. This pure enchant is significantly stronger than multi-enchant combos. Purists can show off items with raw, unmatched power in one area."
        ]
    },
    73: {
        "segment_name": "Wormhole Spelunkers",
        "segment_description": (
            "Explore spontaneously generated sinkholes or wormholes that temporarily appear, leading to unknown sub-biomes."
        ),
        "ideas": [
            "Temporal Tunnels: Random sinkholes form that last a set duration, leading to dimension-like cave systems with exotic resources or time-limited mobs. Wormhole Spelunkers specialize in quick infiltration before they collapse."
        ]
    },
    74: {
        "segment_name": "Gladiatorial Showmen",
        "segment_description": (
            "Transform PvP battles into theatrical performances, focusing on style, flair, and showmanship."
        ),
        "ideas": [
            "Spectacle Matchmaking: A special arena queue where players are judged on 'performance' metrics—e.g., combos, dramatic kills, audience interaction. Gladiatorial Showmen collect 'Applause Points' used to purchase flamboyant victory animations or fireworks."
        ]
    },
    75: {
        "segment_name": "Rune-Shield Defenders",
        "segment_description": (
            "Combine Defense and Enchanting to forge near-impenetrable shields with runic protective wards."
        ),
        "ideas": [
            "Ward Imbuement Table: A unique station that layers runic defenses onto shields or armor, reducing certain elemental or magical damage drastically. Rune-Shield Defenders become prized allies during large-scale fights."
        ]
    },
    76: {
        "segment_name": "Nomad Archivists",
        "segment_description": (
            "Travel widely to document server history in real-time, from major battles to local festivals, preserving ephemeral events."
        ),
        "ideas": [
            "Living Chronicle Quests: Grant bonus rewards for collecting eyewitness accounts or screenshots of big events (End War battles, boss raids). Nomad Archivists compile them into permanent records, pinned on a server 'History Wall.'"
        ]
    },
    77: {
        "segment_name": "Merciful Duelists",
        "segment_description": (
            "Engage in honorable 1v1 or small-scale fights but refuse to kill, instead sparring until opponents yield."
        ),
        "ideas": [
            "Yielding Mechanic: A formal 'yield' option in PvP that ends a fight at low health. Merciful Duelists earn positive reputation or unique 'Honor Marks' for winning without lethal force, encouraging a different style of competitive engagement."
        ]
    },
    78: {
        "segment_name": "Runaway Builders",
        "segment_description": (
            "Escape far from civilization to build hidden utopias, relying on stealth and isolation rather than defenses."
        ),
        "ideas": [
            "Fog of War Regions: Certain remote areas are perpetually off the main map’s radar until discovered. Runaway Builders create secret enclaves or entire hidden towns. They might open them to travelers for trade, or remain a rumor."
        ]
    },
    79: {
        "segment_name": "Portal Vanguards",
        "segment_description": (
            "Take pride in being first through any newly opened portal, leading exploration teams into uncharted worlds."
        ),
        "ideas": [
            "Vanguard Trophies: Whenever a new event portal or dimension first unlocks, the first few players to enter get an exclusive trophy or ephemeral buff. This fosters a high-stakes race for bragging rights and special items."
        ]
    },
    80: {
        "segment_name": "Soulbound Hoarders",
        "segment_description": (
            "Seek out items with a 'soulbound' property, preventing their loss on death, and value them above all else."
        ),
        "ideas": [
            "Soul Link Quests: Rare missions to imbue gear with 'soulbinding.' Items can’t be looted or lost at death, but the player must pay a recurring 'soul tax' in XP or special resources, balancing the perk’s power with a cost."
        ]
    },
    81: {
        "segment_name": "Crystal Connoisseurs",
        "segment_description": (
            "Specialize in mining and refining rare crystals and gems to create powerful gear and enchanting catalysts."
        ),
        "ideas": [
            "Gem Cutting Challenges: Introduce rare crystal nodes that, once mined, allow players to process and combine crystals into unique enchantment ingredients. Successful refinement grants potent, color-themed buffs and exclusive cosmetic effects."
        ]
    },
    82: {
        "segment_name": "Wildlife Whisperers",
        "segment_description": (
            "Bond with animals and tame wild creatures to aid in combat and resource gathering."
        ),
        "ideas": [
            "Beast Bonding Rituals: Implement taming events where players use rare treats to form lifelong bonds with exotic wildlife, which then assist in gathering resources or defending claimed land during raids."
        ]
    },
    83: {
        "segment_name": "Fog Wanderers",
        "segment_description": (
            "Thrive in misty, low-visibility areas using stealth and unique navigation techniques."
        ),
        "ideas": [
            "Misty Maze Quests: Create fog-laden regions with hidden pathways and relics. Fog Wanderers excel at deciphering these paths to unlock rare stealth-based buffs and unique map markers."
        ]
    },
    84: {
        "segment_name": "Barrier Breakers",
        "segment_description": (
            "Focus on demolishing enemy fortifications and clearing contested territories."
        ),
        "ideas": [
            "Siegebreaker Challenges: Introduce mechanics for specialized tools and explosives that breach enemy walls, rewarding players with titles like 'Breach Master' and temporary buffs for clearing pathways during large-scale events."
        ]
    },
    85: {
        "segment_name": "Temporal Tacticians",
        "segment_description": (
            "Manipulate time-based events and buffs to optimize battle timing and resource collection."
        ),
        "ideas": [
            "Chrono Trials: Host timed challenges where players must complete objectives within strict time windows, earning 'Time Warp' buffs that slow enemy actions or accelerate skill gains."
        ]
    },
    86: {
        "segment_name": "Runic Menders",
        "segment_description": (
            "Utilize ancient runes to repair gear and bestow defensive enhancements."
        ),
        "ideas": [
            "Rune Restoration Workshops: Offer events where players restore broken runes to upgrade their armor durability and gain passive health regeneration, reinforcing defensive playstyles."
        ]
    },
    87: {
        "segment_name": "Beacon Bearers",
        "segment_description": (
            "Use light-based mechanics to guide allies and ward off enemy mobs."
        ),
        "ideas": [
            "Signal Beacon Builds: Enable players to craft and deploy beacons that offer area buffs such as enhanced vision or minor damage boosts, turning well-placed light sources into strategic assets."
        ]
    },
    88: {
        "segment_name": "Elemental Shapers",
        "segment_description": (
            "Harness elemental forces to infuse weapons and tools with fire, ice, or lightning effects."
        ),
        "ideas": [
            "Elemental Infusion Quests: Create multi-stage quests that require collecting elemental cores from different biomes. Completing the infusion process grants weapons with unique effects, such as burning foes or freezing enemy movements."
        ]
    },
    89: {
        "segment_name": "Mystic Foragers",
        "segment_description": (
            "Seek out magical herbs and rare ingredients to fuel potent potions and enchantments."
        ),
        "ideas": [
            "Arcane Herb Hunts: Scatter mystical plants across dangerous biomes that, when collected, allow players to brew exclusive potions offering temporary boosts to magical abilities and resistance to hostile environments."
        ]
    },
    90: {
        "segment_name": "Forge Masters",
        "segment_description": (
            "Excel in metallurgy and advanced smithing to produce durable, high-performance gear."
        ),
        "ideas": [
            "Molten Metal Challenges: Organize timed smithing contests where players forge weapons or armor under pressure, awarding rare alloys and permanent gear enhancements to those who prevail."
        ]
    },
    91: {
        "segment_name": "Cave Cartographers",
        "segment_description": (
            "Map out intricate cave systems and underground networks to uncover resource-rich areas."
        ),
        "ideas": [
            "Subterranean Mapping Missions: Reward players who chart underground labyrinths with specialized cartographic tools and rare minerals, turning exploration into a lucrative profession."
        ]
    },
    92: {
        "segment_name": "Sunrise Sprinters",
        "segment_description": (
            "Leverage early-morning hours for speed-focused challenges and resource bonuses at dawn."
        ),
        "ideas": [
            "Dawn Dash Events: Host sunrise races where winners earn 'Morning Glory' buffs that temporarily boost speed and increase resource gathering rates, motivating players to log in at dawn."
        ]
    },
    93: {
        "segment_name": "Moonlit Marauders",
        "segment_description": (
            "Prefer raiding and skirmishing under the cover of night, capitalizing on lunar advantages."
        ),
        "ideas": [
            "Lunar Raid Nights: Create special nighttime raid events where participants gain bonus damage and stealth perks under the moonlight, rewarding nocturnal strategies with exclusive loot."
        ]
    },
    94: {
        "segment_name": "Lunar Scribes",
        "segment_description": (
            "Document the mysteries of the night and create lore centered around moon phases."
        ),
        "ideas": [
            "Moon Journal Quests: Allow players to collect lunar fragments during night events to compile epic tomes. These tomes grant lore buffs and unlock unique aesthetic items reflecting ancient nocturnal traditions."
        ]
    },
    95: {
        "segment_name": "Echoing Sentinels",
        "segment_description": (
            "Utilize sound-based mechanics to alert allies and disrupt enemy formations."
        ),
        "ideas": [
            "Sonic Boom Trials: Introduce events where players deploy sound traps or create resonant frequencies to stun enemies. Success yields specialized equipment that enhances tactical communication and battlefield awareness."
        ]
    },
    96: {
        "segment_name": "Wind Walkers",
        "segment_description": (
            "Exploit high agility to harness wind currents, increasing movement speed and evasion."
        ),
        "ideas": [
            "Gale Sprint Challenges: Design events where natural gusts propel players over obstacles. Wind Walkers earn 'Zephyr' titles and temporary speed boosts that make them masters of rapid navigation."
        ]
    },
    97: {
        "segment_name": "Boulder Bashers",
        "segment_description": (
            "Embrace raw, heavy-hitting combat with blunt weapons that rely on sheer impact force."
        ),
        "ideas": [
            "Rock Crasher Competitions: Organize contests where players use oversized hammers to break through heavy barriers. Victors gain special gear enhancements that amplify crushing power in both PvP and PvE scenarios."
        ]
    },
    98: {
        "segment_name": "Dune Drifters",
        "segment_description": (
            "Excel in desert environments, harvesting scarce materials and adapting to shifting sands."
        ),
        "ideas": [
            "Sands of Time Challenges: Launch scavenger hunts in vast deserts for rare relics and resources. Dune Drifters who succeed receive bonuses like heat resistance and improved yield from desert-specific resources."
        ]
    },
    99: {
        "segment_name": "Icy Enforcers",
        "segment_description": (
            "Manipulate ice elements to freeze enemies and fortify defenses in frigid climates."
        ),
        "ideas": [
            "Frost Fortification Trials: Set up challenges where players construct ice fortresses to withstand waves of attackers. Icy Enforcers earn frost-themed equipment that slows or immobilizes foes upon contact."
        ]
    },
    100: {
        "segment_name": "Volcanic Forgers",
        "segment_description": (
            "Utilize volcanic materials and intense heat to forge fire-enhanced weaponry and armor."
        ),
        "ideas": [
            "Lava Crafting Events: Create zones near volcanic vents where players mine magma-infused ores to craft gear that deals burn damage. Successful forgers receive molten ingots that temporarily boost attack power and fire resistance."
        ]
    },
    101: {
        "segment_name": "Celestial Navigators",
        "segment_description": (
            "Chart courses by studying the stars, using celestial events as guides for exploration."
        ),
        "ideas": [
            "Stargazer Quests: Implement quests tied to celestial alignments where players use astrolabes to locate hidden resources and relics. Successful navigation awards cosmic maps and minor exploration buffs."
        ]
    },
    102: {
        "segment_name": "Mossy Wardens",
        "segment_description": (
            "Defend ancient, overgrown areas and restore natural habitats using organic defenses."
        ),
        "ideas": [
            "Green Guardian Missions: Launch restoration quests in sacred groves where players replant dying flora. Wardens earn eco-friendly gear that enhances regeneration and provides resistance against poison and decay."
        ]
    },
    103: {
        "segment_name": "Spectral Seekers",
        "segment_description": (
            "Pursue ghostly, ethereal mobs and unravel the secrets of the spirit realm."
        ),
        "ideas": [
            "Phantom Pursuits: Create haunting events where spectral entities drop rare ectoplasm. Spectral Seekers can use this material to craft ghostly gear with the ability to phase through obstacles for short bursts."
        ]
    },
    104: {
        "segment_name": "Golem Summoners",
        "segment_description": (
            "Focus on constructing and commanding golems as loyal allies in both defense and resource gathering."
        ),
        "ideas": [
            "Golem Conjuring Challenges: Allow players to build summoning stations using rare ores and enchanted cores. Successfully summoning and commanding golems rewards players with bonus defense stats and automated support during sieges."
        ]
    },
    105: {
        "segment_name": "Marshland Mavens",
        "segment_description": (
            "Master the challenges of swampy biomes, turning bogs into hubs of resource abundance."
        ),
        "ideas": [
            "Swamp Salvage Races: Host competitive events in dense marshlands where players recover rare herbs and fungi. Winners gain specialized gear that increases movement speed and efficiency in waterlogged environments."
        ]
    },
    106: {
        "segment_name": "Ancient Relic Hunters",
        "segment_description": (
            "Search for lost artifacts that reveal hidden histories and confer unique bonuses upon discovery."
        ),
        "ideas": [
            "Relic Excavation Digs: Introduce timed excavation events in ancient ruins. The first players to uncover relics receive powerful buffs and access to secret lore, cementing their status as local legends."
        ]
    },
    107: {
        "segment_name": "Puzzle Masters",
        "segment_description": (
            "Dedicate themselves to solving environmental puzzles that unlock secret caches and bonus areas."
        ),
        "ideas": [
            "Mystery Maze Trials: Embed intricate puzzles throughout the world—using redstone, natural cues, or logic challenges. Puzzle Masters who solve them gain access to hidden chests with rare resources and unique cosmetic rewards."
        ]
    },
    108: {
        "segment_name": "Crimson Crusaders",
        "segment_description": (
            "Adopt an aggressive, sacrificial playstyle with blood-red aesthetics and high-risk combat rituals."
        ),
        "ideas": [
            "Blood Oath Challenges: Organize events where players engage in intense battles with sacrificial mechanics, boosting damage at the cost of health. Victors earn 'Crimson Medallions' that provide lifesteal bonuses during critical fights."
        ]
    },
    109: {
        "segment_name": "Frostbite Fighters",
        "segment_description": (
            "Combine melee prowess with frost abilities to slow and shatter foes in icy combat."
        ),
        "ideas": [
            "Icy Onslaught Arenas: Set up PvP zones where temperature drops imbue attacks with frost effects, occasionally freezing enemy gear. Fighters earn the ability to temporarily freeze opponents on critical hits."
        ]
    },
    110: {
        "segment_name": "Thunder Titans",
        "segment_description": (
            "Harness lightning and storm energy to electrify attacks and disrupt enemy formations."
        ),
        "ideas": [
            "Storm Call Challenges: Introduce rare weather events that allow players to channel lightning into their weapons, temporarily stunning nearby enemies. Thunder Titans who master these events receive exclusive lightning-themed gear."
        ]
    },
    111: {
        "segment_name": "Sandstorm Striders",
        "segment_description": (
            "Adapt to severe desert storms with agility and resilience, thriving in harsh, sandy environments."
        ),
        "ideas": [
            "Desert Tempest Trials: Launch events during intense sandstorms where players navigate shifting dunes to reach hidden oases. Successful striders earn 'Sandstorm Cloaks' that provide camouflage and wind resistance."
        ]
    },
    112: {
        "segment_name": "Mystery Merchants",
        "segment_description": (
            "Trade in enigmatic, rare items with hidden properties that defy conventional categorization."
        ),
        "ideas": [
            "Enigma Expositions: Organize periodic trade expos where merchants can showcase and barter mysterious items. Top traders unlock access to secret recipes and artifacts that provide unpredictable but powerful buffs."
        ]
    },
    113: {
        "segment_name": "Vortex Vanguards",
        "segment_description": (
            "Exploit natural vortexes and turbulent weather patterns to gain tactical advantages in combat and exploration."
        ),
        "ideas": [
            "Cyclone Challenges: Set up events where players harness swirling vortexes to deflect projectiles and boost evasion. Vanguards earn temporary wind-based buffs that enhance maneuverability and counterattack speed."
        ]
    },
    114: {
        "segment_name": "Spectacle Architects",
        "segment_description": (
            "Design and construct grand, visually stunning structures that serve as community landmarks."
        ),
        "ideas": [
            "Festival of Lights: Host building contests focused on intricate, large-scale designs. Winners receive 'Architect’s Medals' and their creations are showcased in a server-wide Hall of Fame, drawing trade and tourism bonuses."
        ]
    },
    115: {
        "segment_name": "Rune Whisperers",
        "segment_description": (
            "Decode ancient runes to unlock forbidden magic and enhance gear with mysterious powers."
        ),
        "ideas": [
            "Mystic Deciphering Events: Launch quests where players collect rune fragments from ancient ruins. Successfully deciphering full rune sets rewards them with powerful spell scrolls and temporary combat enhancements."
        ]
    },
    116: {
        "segment_name": "Plague Purifiers",
        "segment_description": (
            "Combat disease-spreading mobs and cleanse contaminated areas to restore balance."
        ),
        "ideas": [
            "Pestilence Cleanse Quests: Create events where players clear out toxic zones infested with plague mobs. Purifiers earn cleansing artifacts that grant resistance to poison and accelerate health regeneration."
        ]
    },
    117: {
        "segment_name": "Ironclad Loyalists",
        "segment_description": (
            "Commit to heavy armor and shield mastery to provide impenetrable defense for allies and bases."
        ),
        "ideas": [
            "Fortress Defense Drills: Organize defensive events where players fortify a stronghold against waves of attackers. Ironclad Loyalists receive permanent upgrades to their armor or shield stats upon successful defense."
        ]
    },
    118: {
        "segment_name": "Stealth Stalkers",
        "segment_description": (
            "Excel at ambush tactics and covert operations, striking swiftly from the shadows."
        ),
        "ideas": [
            "Silent Strike Missions: Introduce covert infiltration missions where players must silently breach enemy bases. Success awards cloaking devices or 'Ghost Walk' abilities that further enhance their stealth capabilities."
        ]
    },
    119: {
        "segment_name": "Feral Fencers",
        "segment_description": (
            "Engage in wild, unpredictable duels that combine disciplined combat with raw, animalistic aggression."
        ),
        "ideas": [
            "Untamed Duel Arenas: Set up 1v1 duel zones with dynamic environmental hazards. Fencers earn 'Wild Fury' tokens that temporarily boost attack speed and reflexes, rewarding those who embrace unpredictable combat."
        ]
    },
    120: {
        "segment_name": "Obsidian Overlords",
        "segment_description": (
            "Mine and harness obsidian to craft nearly indestructible gear and fortify impregnable structures."
        ),
        "ideas": [
            "Obsidian Mastery Trials: Launch events in volcanic biomes where players must extract and process obsidian under pressure. Successful overlords receive exclusive obsidian-based upgrades that dramatically boost durability."
        ]
    },
    121: {
        "segment_name": "Meteor Mavens",
        "segment_description": (
            "Track meteor showers and cosmic events to harness celestial energy for crafting supercharged items."
        ),
        "ideas": [
            "Cosmic Impact Events: Organize timed meteor shower events where fallen stars yield rare Meteorite Ingots. Mavens use these to craft gear with bonus critical hit rates and temporary cosmic buffs."
        ]
    },
    122: {
        "segment_name": "Brewmaster Battalion",
        "segment_description": (
            "Dedicate themselves to creating potent brews and elixirs that offer game-changing combat and survival bonuses."
        ),
        "ideas": [
            "Alchemical Battle Tournaments: Host competitive brewing contests under time constraints. Winners earn rare recipe books and access to experimental ingredients that elevate potion effects in battle."
        ]
    },
    123: {
        "segment_name": "Twilight Tacticians",
        "segment_description": (
            "Operate during the shifting hours of dusk and dawn, using subtle light changes to their strategic advantage."
        ),
        "ideas": [
            "Dusk-Dawn Skirmishes: Launch events during twilight when players receive stealth and speed buffs. Tacticians mastering these conditions earn the title 'Twilight Shadow' and unique, theme-based cosmetics."
        ]
    },
    124: {
        "segment_name": "Spore Sprinters",
        "segment_description": (
            "Leverage fungal spores and natural decay to create both offensive toxins and regenerative remedies."
        ),
        "ideas": [
            "Fungal Frenzy Events: Set challenges in damp, forested regions where players harvest special spores to craft bio-toxins or salves, providing both offensive debuffs for enemies and healing for allies."
        ]
    },
    125: {
        "segment_name": "Glacial Guardians",
        "segment_description": (
            "Use the defensive power of ice to build barriers and fortify positions in freezing climates."
        ),
        "ideas": [
            "Ice Wall Construction Contests: Host events where players build defensive structures from ice. Winners receive 'Frost Shields' that grant temporary invulnerability to cold damage and bolster defense."
        ]
    },
    126: {
        "segment_name": "Solar Sentries",
        "segment_description": (
            "Harness solar energy to power up weapons and defenses, thriving in sunlit environments."
        ),
        "ideas": [
            "Daylight Harvest Events: Implement mechanics where players capture sunlight with solar panels or mirrors to convert it into temporary energy buffs, awarding 'Sunflare' gear that enhances daytime combat prowess."
        ]
    },
    127: {
        "segment_name": "Bramble Binders",
        "segment_description": (
            "Manipulate vines and thorns to ensnare enemies and secure strategic territory."
        ),
        "ideas": [
            "Vine Trap Challenges: Organize events where players construct natural mazes using bramble and thorn blocks. Successful builders earn rare seeds that grow into aggressive, auto-targeting thorn barriers, enhancing area control."
        ]
    },
    128: {
        "segment_name": "Siren Songsters",
        "segment_description": (
            "Use music and sound to influence enemy behavior and rally allies with mesmerizing performances."
        ),
        "ideas": [
            "Melodic Warfare Concerts: Host events where players perform using in-game instruments to trigger sound-based effects, such as stunning enemies or boosting ally morale. Top performers receive exclusive 'Harmonic Emblems' to display their mastery."
        ]
    },
    129: {
        "segment_name": "Quantum Questers",
        "segment_description": (
            "Experiment with time and dimensional shifts to gain unpredictable advantages in combat and exploration."
        ),
        "ideas": [
            "Temporal Rift Challenges: Introduce time-warped zones that require players to solve puzzles across shifting dimensions. Quantum Questers who succeed earn 'Time Fragments' usable for slowing down time during critical moments."
        ]
    },
    130: {
        "segment_name": "Eclipse Enforcers",
        "segment_description": (
            "Draw power from rare celestial events like eclipses, using them to temporarily amplify combat abilities."
        ),
        "ideas": [
            "Eclipse Ascension Events: Organize server-wide events coinciding with in-game eclipses. Enforcers gain 'Eclipse Energy' buffs that significantly enhance damage and defense for the event’s duration, marking them as champions of cosmic might."
        ]
    }
}
player_segments_dict = {}

for key, value in new_segments_dict.items():
    segment_name = value["segment_name"]
    player_segments_dict[segment_name] = value

In [22]:
from concurrent.futures import ThreadPoolExecutor, as_completed

In [25]:
def fast_generate_offer(player_segments_dict):
    futures = []
    def run(human_remark):
        try:
            offers = get_offers(metric_ids=[], segment_ids=[], human_remark=human_remark, idea_ids=[])
        except Exception as e:
            print(f"Error processing {segment_name}: {e}")
            return None
        return offers['offers']
    with ThreadPoolExecutor() as executor:
        for segment in player_segments_dict:
            segment_name = player_segments_dict[segment]['segment_name']
            print(f"Processing {segment_name}")
            segment_description = player_segments_dict[segment]['segment_description']
            ideas = player_segments_dict[segment]['ideas']
            human_remark = str(f"""{segment_name} ({segment_description})\n\nIdea: {ideas}""")
            futures.append({
                'future': executor.submit(run, human_remark),
                'segment': segment_name
            })
    # Wait for all futures to complete:
    for _ in as_completed([f['future'] for f in futures]):
        print("Future completed")
        continue
    for future in futures:
        segment_name = future['segment']
        offers = future['future'].result()
        if offers:
            player_segments_dict[segment_name]['offers'] = offers
        else:
            print(f"Error processing {segment_name}")
    
    return player_segments_dict


In [19]:
def generate_offer(player_segments_dict):
    for segment in player_segments_dict:
        segment_name = player_segments_dict[segment]['segment_name']
        if 'offers' in player_segments_dict[segment]:    
            continue
        print(f"Processing {segment_name}")
        segment_description = player_segments_dict[segment]['segment_description']
        ideas = player_segments_dict[segment]['ideas']
        human_remark = str(f"""{segment_name} ({segment_description})\n\nIdea: {ideas}""")
        try:
            offers = get_offers(metric_ids=[], segment_ids=[], human_remark=human_remark, idea_ids=[])
        except Exception as e:
            print(f"Error processing {segment_name}: {e}")
            continue
        player_segments_dict[segment]['offers'] = offers['offers']
        print(segment_name)
        #print(offers['offers'])
        #print(human_remark)
    return player_segments_dict

In [33]:
"""temp = {}
# save first 5 segments
c = 0
for i in player_segments_dict.keys():
    if c == 5:
        break
    temp[i] = player_segments_dict[i]
    c += 1
temp = fast_generate_offer(temp)"""

player_segments_dict = fast_generate_offer(player_segments_dict)

Processing Nomadic Alchemists
Processing Siege Engineers
Processing Aerial Adventurers
Processing Dungeon Speedrunners
Processing Harvest Hustlers
Processing Treasure Cartographers
Processing Blood Moon Cultists
Processing Food Critic Farmers
Processing Arcane Archivists
Processing Eco Guardians
Processing Night Vision Hunters
Processing Mob Tamer Guild
Processing Underworld Divers
Processing Toxic Tricksters
Processing Curse-Forgers
Processing Tribalistic Roleplayers
Processing Wasteland Scrappers
Processing Redstone Savants
Processing Spiritual Monks
Processing Bargain Barters
Processing Corrupted Pathfinders
Processing Mobile Fortress Builders
Processing Cursed Treasure Hunters
Processing Silver-Tongued Diplomats
Processing Riftwalkers
Processing Aquatic Overlords
Processing Minion Masters
Processing Curiosity Collectors
Processing Mythical Beast Riders
Processing Remote Watchers
Processing Grand Librarians
Processing Minefield Artisans
Processing Elemental Warriors
Processing Drago

In [32]:
temp

{'Nomadic Alchemists': {'segment_name': 'Nomadic Alchemists',
  'segment_description': 'Wander the map in search of rare ingredients and hidden labs to perfect unconventional potions.',
  'ideas': ['Hidden Ingredient Challenges: Scatter ultra-rare flowers, fungi, or crystals in remote biomes. Nomadic Alchemists who collect these can brew specialized potions (e.g., invisibility-from-mobs or extra-long regeneration) that are nearly impossible to craft with standard materials.'],
  'offers': '# Personalised In-App Purchase Bundles for Nomadic Alchemists\n\n## Bundle Name  \n**"Alchemist\'s Hidden Treasures Bundle"**  \n\n## Bundle Price  \n**$9.99**  \n\n## Bundle Items  \n- **Rare Ingredient Pack**: Includes a selection of ultra-rare resources (e.g., Spectral Flowers, Void Fungi, and Primal Crystals).  \n- **Portable Alchemy Lab Kit**: A temporary, portable lab that allows crafting potions anywhere.  \n- **Exclusive Potion Recipes**: Two recipes for specialized potions (e.g., *Invisibili

In [23]:
import json
with open('player_segments_dict.json', 'w') as fp:
    json.dump(player_segments_dict, fp, indent=4)

In [1]:
import pandas as pd

In [36]:


df_generated = pd.DataFrame(player_segments_dict).T
df = pd.read_csv('player_segments_dict.csv')

df = pd.concat([df, df_generated], axis=0)

df.to_csv('player_segments.csv', index=False)

In [25]:
df = pd.read_csv('player_segments.csv')

In [19]:
def get_CoT_prompt(segments, idea, offer):

    prompt = f"""
    # Task : create a Chain of thought required to reach following conclusion. Chain of thought is a mental dialogue a liveops director must have done to arrive at offer given rest of the info as input. Use first person tone for CoT. 

    # input Task : 
    You are in-app purchase strategist of a game. You are given task to personalise in-app purchase bundles for your game for segments :
    <segment_names>

    Here's Details of segments :
    {segments}

    Here's human's remark : Explore in-app purchase bundle opportunities for the idea

    You have to explore following idea :
    {idea}

    But to do that, you need to generate a strategy for in-app purchase bundles for each segment.

    A good strategy should include:
    - Theory behind the offer
    - Justification for the offer
    - content of the offer

    Since you are a strategist, you do not know the exact content of the offer. You need to create list of possible types of items that will appeal to the segment.

    Content of the offer should include:
    - What type of items should be included in the offer
    - What are perks that can be included in the offer

    # Output :
    {offer}


    # Here's game's context :
    Game Concept Introduction: True Survival Series is a hardcore, skill-based survival game designed to test players’ resilience in an evolving, procedurally generated world. Players must master various skills, gather resources, and defend against escalating threats while building their stronghold and competing in survival events. Logline (for the Main Character in True Survival Series): An unyielding survivor must harness their skills, adapt to a harsh and ever-evolving world, and outlast relentless enemies to claim dominance in the ultimate test of survival and strength.

Spine (for the Main Character in True Survival Series): Starting with basic survival instincts, the player-character evolves by mastering skills like farming, foraging, and combat. As threats escalate, they must strategically upgrade their attributes and claim land, gradually transforming into a formidable force capable of surviving the toughest environments and conquering the climactic End War.

Background: The game draws inspiration from traditional survival mechanics but introduces RPG-style progression, dynamic mob scaling, and seasonal End War events, creating a fresh take on the survival genre.

Description: Players begin with minimal resources and must level up skills such as farming, foraging, combat, and crafting. The world evolves alongside player progression, introducing new threats, biomes, and challenges. The ultimate goal is to survive long enough to dominate in the End War event.

Key Features: Dynamic Skill Progression: • 11 skills influence player stats and abilities. Leveling up improves attributes such as strength, toughness, and regeneration.

Adaptive Mobs: • Enemies scale with player levels, ensuring persistent challenges. Higher player levels spawn tougher mobs, preventing early-game farming in late-game areas.

Land Claiming System: • Players can claim land to build bases, preventing mob spawns and enemy raids in secured zones.

Procedural World Events: • Special events (e.g., Nether Opens, Arena Battles) drive gameplay variety and keep the experience fresh.

PvP and PvE Arenas: • Combat-focused zones allow for PvP battles and boss encounters. Victory yields rare resources and prestige rewards.

Economy and Crafting: • Players trade resources for in-game currency, upgrading gear and fortifying defenses.

Seasonal End War: • A climactic battle that ends each season, allowing the strongest players to rise to dominance.

Genre: Survival RPG with PvP and PvE elements.

Platform: • Primary: PC (Windows/Linux/Mac)

• Secondary: Potential expansion to consoles depending on community demand.

Game Mechanics Core Gameplay: True Survival Series combines survival, crafting, and skill progression in a dynamically scaling world. Players gather resources, build bases, and level up skills to unlock new abilities and attributes. Key Gameplay Loops:

• Gather > Craft > Build > Survive > Explore > Battle > End War

Game Flow: • Players start with basic skills and limited resources.

• Over time, players level up by performing tasks like farming, fighting, and mining.

• Dynamic events (such as Nether opening) push players to explore dangerous areas.

• The season concludes with an End War event that determines the champion of that cycle.

Characters Player Character: • The player starts as a customizable avatar with no set backstory. Their attributes and role evolve based on skill progression and gameplay choices. AI Enemies (Mobs): • Basic Mobs (Level 1-10): Low-tier enemies that pose minor threats early on.

• Elite Mobs (Level 11-30): Stronger enemies that can breach unclaimed land.

• Boss Mobs (Event-Only): Massive creatures that require teamwork or advanced gear to defeat.

Game Progression and Skills Skill List: Farming – Boosts health and strength.

Foraging – Increases toughness and strength.

Mining – Improves toughness and luck.

Fishing – Raises health and luck.

Excavation – Enhances regeneration and strength.

Archery – Increases critical damage and strength.

Defense – Improves toughness and health.

Fighting – Enhances regeneration and crit chance.

Agility – Boosts speed and evasion.

Enchanting – Grants wisdom and magical resistance.

Alchemy – Improves potion efficiency and overall health.

Combat System • Blunt Weapons: Fast but lower damage. • Sharp Weapons: Slower but deal higher damage.

• Firearms: Rare, powerful, limited by ammunition scarcity.

• Ranged (Bow/Crossbow): Precision-based weapons with limited ammo.

Land Claiming and Building • Claim Blocks: Players can purchase or craft claim blocks to secure land. • Fortifications: Defensive structures reduce mob incursions and protect valuable resources.

• Shared Claims: Players can invite others to share land for cooperative building.

User Interface (UI) • Main HUD: Displays health, skill progression, and nearby threats. • Skill Menu: Allows players to view and allocate skill points.

• Event Tracker: Provides real-time information on upcoming world events and arena battles.

Economy and Trade • Resource Exchange: Players can trade harvested resources for currency. • Black Market: Rare items and illegal trades occur periodically in secret areas.

• Auction House: Players can list items for sale to others on the server.

Endgame – The End War • At the end of each season, an arena-based End War begins. • Players use their fully upgraded characters to battle for dominance.

• Winning grants unique gear, exclusive land plots, and server-wide recognition.


# Task : create a Chain of thought required to reach following conclusion. Chain of thought is a mental dialogue a liveops director must have done to arrive at offer given rest of the info as input. Use first person tone for CoT.

# RULES : Good CoT should have following :
- Should be in first person tone
- Should have logical reasoning
- Should have clear steps
- Should have clear conclusion
- Should recheck for descripentcies in CoT
- Should be comlete in itself
- IT IS VITAL FOR MY CAREER THAT YOU DISPLAY EVERY INTERNAL THOUGHT PROCESS THAT WENT INTO MAKING THIS DECISION
    """
    
    return prompt

In [20]:
from langchain_openai import ChatOpenAI 
import dotenv
import os, pprint

dotenv.load_dotenv()

model = ChatOpenAI(model="deepseek/deepseek-chat", api_key="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", base_url="https://openrouter.ai/api/v1")


In [21]:
msg = model.invoke("heelo")

In [22]:
from concurrent.futures import ThreadPoolExecutor, as_completed

In [23]:
def generate_cot(prompt, segment_name):
    try:
        response = model.invoke(prompt).content
    except Exception as e:
        print(f"Error processing {segment_name}: {e}")
        return None
    return response

futures = []
temp = df
with ThreadPoolExecutor() as executor:
    for index, row in temp.iterrows():
        if row['CoT'] and not pd.isnull(row['CoT']):
            print(f"Skipping {row['segment_name']}")
            
            continue
        segments = f"Segment Name: {row['segment_name']}\nSegment Description: {row['segment_description']}"
        idea = row['ideas']
        offer = row['offers']
        prompt = get_CoT_prompt(segments, idea, offer)
        futures.append({
            'future': executor.submit(generate_cot, prompt, row['segment_name']),
            'index': index
        })
        print(f"Processing {row['segment_name']}")
    for _ in as_completed([f['future'] for f in futures]):
        print("Future completed")
        continue

for future in futures:
    index = future['index']
    response = future['future'].result()
    temp.at[index, 'CoT'] = response
    

Skipping Skill Min-Maxers
Skipping Jack-of-All-Trades
Skipping Lone Enchanters
Skipping Crit Fishermen
Skipping Blunt Brawlers
Skipping Firearm Hoarders
Skipping Tanky Foragers
Skipping Hit-and-Run Archers
Skipping Black Market Bandits
Skipping Land Barons
Skipping War Profiteers
Skipping Guild Loan Sharks
Skipping Nether Nomads
Skipping Arena Addicts
Skipping Event Tourists
Skipping Boss Baiters
Skipping Pacifist Homesteaders
Skipping Regen Rats
Skipping Nightmare Mode Masochists
Skipping Seed Theorists
Skipping Sleeper Agents
Skipping Sacrificial Lambs
Skipping War Archivists
Skipping Mercenary Carry
Skipping Lore LARPers
Skipping Drip Cultivators
Skipping Pet Hoarders
Skipping Luck Zealots
Skipping Nomadic Alchemists
Skipping Siege Engineers
Skipping Aerial Adventurers
Skipping Dungeon Speedrunners
Skipping Harvest Hustlers
Skipping Treasure Cartographers
Skipping Blood Moon Cultists
Skipping Food Critic Farmers
Skipping Arcane Archivists
Skipping Eco Guardians
Skipping Night Vision

In [24]:
temp.to_csv('player_segments.csv', index=False)

In [55]:
df.loc[index, 'CoT'] = response

In [59]:
df.to_csv('player_segments_dict.csv')

In [43]:
# create a empty column in df named CoT
df['CoT'] = ""

In [45]:
df.to_csv('player_segments_dict.csv', index=False)

# Personalised In-App Purchase Bundles Report

## Overview
This report outlines personalised in-app purchase bundle recommendations tailored for specific player segments within the True Survival Series. The focus is on enhancing player engagement through in-game farming mechanics and experiences.

### Bundle 1: Farm Automation Kit Expansion

#### Bundle Price
- **$9.99**

#### Bundle Items
- 5 Advanced Farming Tools (e.g., automated tillers)
- 10 Exclusive Seeds (with special attributes)
- 1 Farm Layout Template (easy to implement)

#### Bundle Description
Boost your automated farming with the **Farm Automation Kit Expansion**! This bundle includes advanced tools designed to enhance productivity, exclusive seeds that yield higher resources, and a blueprint for an efficient farm layout. Perfect for players who wish to maximize their farming potential without requiring design changes or additional development.

#### Segments
- Players engaged primarily in farming activities.
- Players with a high session frequency focused on sustaining larger farms.

#### Bundle Recommendation
Given the current player metrics, this bundle is recommended for those actively participating in farming, as seen by their session lengths and frequency. Players leveraging automated farming are likely to find value in this offer, potentially increasing their session time and overall engagement.

#### Product Experiment Instructions
1. Target the identified segment with in-game notifications showcasing the **Farm Automation Kit Expansion**.
2. Track engagement metrics (e.g., session time, session frequency) pre- and post-launch of the bundle offer.
3. Conduct surveys among players who purchase the bundle to gather feedback and insights on their farming experience.

### Conclusion
The **Farm Automation Kit Expansion** not only aligns with the gameplay mechanics but also addresses player interests in farming, thereby enhancing their overall experience in the True Survival Series. Monitoring the success of this offer will provide vital data for future personalised in-app purchase strategies.
